### ISSUE TITLE
Default position of colorbar labels overlays its values

### ISSUE THREAD

**Issue by dokato** [matplotlib/matplotlib#19029]
### Bug report

**Bug summary**

The default colourbar label is shifted such that it doesn't overlap with the first item of the colorbar ticks. But when negative numbers are available below it overlaps them as shown below.

**Code for reproduction**

```python
cbar = plt.colorbar(im, ax = axes[1, 1])
cbar.ax.set_ylabel('RT(1ax) - RT(2ax) [ms]', rotation=270)
```

**Actual outcome**

<img width="206" alt="Screenshot 2020-11-27 at 09 23 09" src="https://user-images.githubusercontent.com/4547289/100436579-73725600-3097-11eb-9c22-a45d536c4c5b.png">

**Expected outcome**

Such that label of colourbar is shifted more to the right and it doesn't overlap with negative values from the colorbar axis.

**Matplotlib version**

  * Operating system: MacOS
  * Matplotlib version: 3.2.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): MacOSX
  * Python version: 3.7.7

  * Other libraries: installed from conda

**Comment by jklymak** [comment]
Is that true for all ylabels?

**Comment by timhoffm** [comment]
`rotation=270` is the culprit. Minimal example:
~~~
fig, axs = plt.subplots(2)
for ax in axs:
    ax.plot([1, 3, 2])
    ax.yaxis.set_label_position("right")
    ax.yaxis.tick_right()
    if ax is axs[0]:
        ax.set_ylabel('RT(1ax) - RT(2ax) [ms]')
    else:
        ax.set_ylabel('RT(1ax) - RT(2ax) [ms]', rotation=270)
plt.show()
~~~
![grafik](https://user-images.githubusercontent.com/2836374/100483319-c5979380-30f8-11eb-971f-556f36cf2268.png)

**Comment by jklymak** [comment]
Good catch - I didn't see the 270.  Can you just fix by changing the vertical (or maybe horizontal) alignment?

**Comment by timhoffm** [comment]
Indeed, `verticalalignment='baseline'` fixes it:

![grafik](https://user-images.githubusercontent.com/2836374/100486408-06e17080-3104-11eb-8b44-e95f791171ba.png)

**Comment by timhoffm** [comment]
The underlying issue is an adjustment of the vertical alignment in `YAxis.set_label_position()`:

https://github.com/matplotlib/matplotlib/blob/5960b5a5e2826c4cbdc0a07e6420daed04971eb2/lib/matplotlib/axis.py#L2369

This assumes 90° rotated y labels. The problem is, that in both my example and the colorbar example the label rotation is still 90° when `set_label_position()` is called. It's only later rotated. So it's not even sufficient to make `set_label_position()` take the rotation into account. That would need to be done in `YAxis._update_label_position`.

More generally, I'm unclear if we should mess at all with the vertical alignment, and if so, if that should rather be done in `_update_label_position()`.

**Comment by jklymak** [comment]
I guess `_update_label_postion` could simply check the extent of the label as well as all the tick labels and make sure its not overlapping?

**Comment by timhoffm** [comment]
> I guess `_update_label_postion` could simply check the extent of the label as well as all the tick labels and make sure its not overlapping?

That would essentially render the vertical alignment ineffective. Maybe we exactly want that? However I suspect such an algorithm might slightly the label position.

**Comment by jklymak** [comment]
I think the vertical alignment is currently basically useless for the automatic placement of x and ylabels (but not if the user specifies y or x), so I don't see any issue making it more useless.  But changing the algorithm *will* change many image tests.
